Ref.: http://lists.busybox.net/pipermail/busybox/2011-March/075325.html

The latest busybox fdisk utility with GPT enabled displays the wrong
size for the disk and partitions.  This appears to be because
smart_ulltoa5() is expecting bytes, where gpt_list_table() is sending
sectors.  The attached patch fixes it for me.

-Justin Maggard

diff --git a/util-linux/fdisk.c b/util-linux/fdisk.c
--- a/util-linux/fdisk.c
+++ b/util-linux/fdisk.c
@@ -423,8 +423,8 @@ static sector_t bb_BLKGETSIZE_sectors(int fd)
 	unsigned long longsectors;
 
 	if (ioctl(fd, BLKGETSIZE64, &v64) == 0) {
-		/* Got bytes, convert to 512 byte sectors */
-		v64 >>= 9;
+		/* Got bytes, convert to sectors */
+		v64 /= sector_size;
 		if (v64 != (sector_t)v64) {
  ret_trunc:
 			/* Not only DOS, but all other partition tables
@@ -1313,10 +1313,7 @@ get_geometry(void)
 static void
 get_geometry(void)
 {
-	int sec_fac;
-
 	get_sectorsize();
-	sec_fac = sector_size / 512;
 #if ENABLE_FEATURE_SUN_LABEL
 	guess_device_type();
 #endif
@@ -1339,7 +1336,7 @@ get_geometry(void)
 	if (dos_compatible_flag)
 		sector_offset = g_sectors;
 
-	g_cylinders = total_number_of_sectors / (g_heads * g_sectors * sec_fac);
+	g_cylinders = total_number_of_sectors / g_heads / g_sectors;
 	if (!g_cylinders)
 		g_cylinders = user_cylinders;
 }
@@ -1939,7 +1939,7 @@ list_disk_geometry(void)
 static void
 list_disk_geometry(void)
 {
-	ullong bytes = ((ullong)total_number_of_sectors << 9);
+	ullong bytes = ((ullong)total_number_of_sectors * sector_size);
 	long megabytes = bytes / 1000000;
 
 	if (megabytes < 10000)
@@ -1952,7 +1952,7 @@ list_disk_geometry(void)
 		   g_heads, g_sectors, g_cylinders);
 	if (units_per_sector == 1)
 		printf(", total %"SECT_FMT"u sectors",
-			total_number_of_sectors / (sector_size/512));
+			total_number_of_sectors);
 	printf("\nUnits = %s of %u * %u = %u bytes\n\n",
 		str_units(PLURAL),
 		units_per_sector, sector_size, units_per_sector * sector_size);
-- 
